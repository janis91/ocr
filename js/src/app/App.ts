import { Util } from './util/Util';
import { OcaService } from './service/OcaService';
import { Controller } from './controller/Controller';
import { View } from './view/View';
import * as handlebarsDropdownTemplate from './view/templates/ocr.hbs';
import { TesseractService } from './service/TesseractService';
import { OC, OCA } from 'global-oc-types';
import { PDFJSStatic } from 'pdfjs-dist';
import { PdfService } from './service/PdfService';
import { PDFDocumentFactory, PDFDocumentWriter } from 'pdf-lib';


declare var OC: OC;
declare var OCA: OCA;
declare var PDFJS: PDFJSStatic;
declare var PDFLib: { PDFDocumentFactory: PDFDocumentFactory, PDFDocumentWriter: PDFDocumentWriter };

/**
 * Nextcloud - OCR
 *
 * This file is licensed under the Affero General Public License version 3 or
 * later. See the COPYING file.
 *
 * @author Janis Koehr <janiskoehr@icloud.com>
 * @copyright Janis Koehr 2019
 */
export class App {
    public util: Util;
    public view: View;
    public ocaService: OcaService;
    public tesseractService: TesseractService;
    public pdfService: PdfService;
    public controller: Controller;
    public initCounter: number = 0;

    constructor() {
        const interval = setInterval(() => {
            if (Util.isDefinedIn('t', window) && Util.isDefinedIn('n', window) &&
                OcaService.checkOCAvailability() && TesseractService.checkTesseractAvailability() &&
                PdfService.checkPdfUtilAvailability()) {
                this.util = new Util();
                this.view = new View(OC.Notification, handlebarsDropdownTemplate, document);
                this.ocaService = new OcaService(OC, OCA);
                this.tesseractService = new TesseractService(document);
                this.pdfService = new PdfService(PDFJS, PDFLib, document);
                this.controller = new Controller(this.util, this.view, this.tesseractService, this.ocaService, this.pdfService, document);
                try {
                    this.controller.init();
                } catch (e) {
                    console.error(e);
                    this.view.displayError(e.message);
                }
                clearInterval(interval);
            }
            if (this.initCounter === 50) {
                console.error('OCR could not be initiallized. Some of the required resources (OC, OCA, Tesseract, etc.) did not load in time.');
                clearInterval(interval);
            }
            this.initCounter++;
        }, 100);
    }
}

export const $app = new App();
