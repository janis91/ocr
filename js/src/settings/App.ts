import { Controller } from './controller/Controller';
import { View } from './view/View';
import * as handlebarsSettingsTemplate from './view/templates/ocr.hbs';
import { HttpService } from './service/HttpService';
import { OC, OCA, OCSingleTranslation, OCMultiTranslation } from '../global-oc-types';
import { Util } from './util/Util';
import axios from 'axios';

declare var OC: OC;
declare var OCA: OCA;
declare var t: OCSingleTranslation;
declare var n: OCMultiTranslation;

/**
 * Nextcloud - OCR
 *
 * This file is licensed under the Affero General Public License version 3 or
 * later. See the COPYING file.
 *
 * @author Janis Koehr <janiskoehr@icloud.com>
 * @copyright Janis Koehr 2019
 */
export class App {
    public view: View;
    public httpService: HttpService;
    public controller: Controller;
    public initCounter: number = 0;

    constructor() {
        const interval = setInterval(() => {
            if (Util.isDefinedIn('t', window) && typeof t === 'function' && Util.isDefinedIn('n', window) && typeof n === 'function'
                && Util.isDefinedIn('OC', window) && Util.isDefinedIn('Notification', (window as any).OC)) {
                this.view = new View(OC.Notification, handlebarsSettingsTemplate, document);
                this.httpService = new HttpService(OC, axios);
                this.controller = new Controller(this.view, this.httpService, document);
                this.controller.init()
                    .catch((e) => {
                        console.error(e);
                        this.view.displayError(e.message);
                    });
                clearInterval(interval);
            }
            if (this.initCounter === 50) {
                console.error('OCR could not be initiallized. Some of the required resources (OC, etc.) did not load in time.');
                clearInterval(interval);
            }
            this.initCounter++;
        }, 100);
    }
}

export const $app = new App();
