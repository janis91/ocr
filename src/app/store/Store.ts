import { generateUrl } from '@nextcloud/router'
import axios from '@nextcloud/axios'
import { AxiosResponse } from 'axios'
import { NextcloudGuiApiService } from '@a/services/NextcloudGuiApiService'
import { Translations } from '@a/configuration/Translations'
import { Util } from '@/common/Util'
import { StoreOptions } from 'vuex'

export const SHOW_MODAL = 'SHOW_MODAL'
export const HIDE_MODAL = 'HIDE_MODAL'
export const LOAD_FAVORITE_LANGUAGES = 'LOAD_FAVORITE_LANGUAGES'
const SET_FAVORITE_LANGUAGES = 'SET_FAVORITE_LANGUAGES'
export const UPDATE_SELECTED_LANGUAGES = 'UPDATE_SELECTED_LANGUAGES'
export const CLEAR_SELECTED_LANGUAGES = 'CLEAR_SELECTED_LANGUAGES'
export const SET_PROCESSING = 'SET_PROCESSING'
const SET_FINISHED = 'SET_FINISHED'
export const STOP_PROCESS_AND_RESET_MODAL_STATE = 'STOP_PROCESS_AND_RESET_MODAL_STATE'
export const SET_SELECTED_FILES = 'SET_SELECTED_FILES'
export const CLEAR_SELECTED_FILES = 'CLEAR_SELECTED_FILES'

export interface State {
  selectedFiles: OCAFile[];
  favoriteLanguages: Map<string, string>;
  showModal: boolean;
  selectedLanguages: Map<string, string>;
  processing: boolean;
}

export default {
  strict: true,
  state: {
    selectedFiles: [] as OCAFile[],
    favoriteLanguages: new Map<string, string>(),
    showModal: false as boolean,
    selectedLanguages: new Map<string, string>(),
    processing: false as boolean
  },
  getters: {
    selectedOptions(state) {
      return Util.mapToOptions(state.selectedLanguages)
    }
  },
  mutations: {
    [SHOW_MODAL](state) {
      state.showModal = true
    },
    [HIDE_MODAL](state) {
      state.showModal = false
    },
    [SET_PROCESSING](state) {
      state.processing = true
    },
    [SET_FINISHED](state) {
      state.processing = false
    },
    [SET_SELECTED_FILES](state, payload: {selectedFiles: OCAFile[]}) {
      state.selectedFiles = payload.selectedFiles
    },
    [CLEAR_SELECTED_FILES](state) {
      state.selectedFiles = []
    },
    [SET_FAVORITE_LANGUAGES](state, payload: { favoriteLanguages: Map<string, string> }) {
      state.favoriteLanguages = payload.favoriteLanguages
      state.selectedLanguages = payload.favoriteLanguages
    },
    [UPDATE_SELECTED_LANGUAGES](state, payload: { selectedLanguages: Map<string, string> }) {
      state.selectedLanguages = payload.selectedLanguages
    },
    [CLEAR_SELECTED_LANGUAGES](state) {
      state.selectedLanguages = new Map()
    }
  },
  actions: {
    async [LOAD_FAVORITE_LANGUAGES]({ commit }) {
      try {
        const { data } = await axios.get<any, AxiosResponse<string[]>>(generateUrl('/apps/ocr/api/personal/languages'))
        const favoriteLanguages = Util.mapToCommonLanguages(data)
        commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages })
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error('Unexpected error while loading favorite languages', e)
        NextcloudGuiApiService.displayError(Translations.TRANSLATION_UNEXPECTED_ERROR_LOAD_FAVORITE_LANGUAGES)
      }
    },
    [STOP_PROCESS_AND_RESET_MODAL_STATE]({ commit, state }) {
      commit(SET_FINISHED)
      commit(HIDE_MODAL)
      commit(UPDATE_SELECTED_LANGUAGES, { selectedLanguages: state.favoriteLanguages })
    }
  }
} as StoreOptions<State>
