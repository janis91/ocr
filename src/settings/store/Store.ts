import Vue from 'vue'
import Vuex, { Store } from 'vuex'
import { generateUrl } from '@nextcloud/router'
import axios from '@nextcloud/axios'
import { AxiosResponse } from 'axios'
import { NextcloudGuiApiService } from '@s/services/NextcloudGuiApiService'
import { Translations } from '@s/configuration/Translations'
import { Util, LanguageOption } from '@/common/Util'

Vue.use(Vuex)

export const SAVE_FAVORITE_LANGUAGES = 'SAVE_FAVORITE_LANGUAGES'
export const LOAD_FAVORITE_LANGUAGES = 'LOAD_FAVORITE_LANGUAGES'
export const SET_FAVORITE_LANGUAGES = 'SET_FAVORITE_LANGUAGES'
export const CLEAR_FAVORITE_LANGUAGES = 'CLEAR_FAVORITE_LANGUAGES'
const SET_STATUS_LOAD = 'SET_STATUS_LOAD'
const SET_STATUS_READY = 'SET_STATUS_READY'

export default new Store({
  strict: true,
  state: {
    favoriteLanguages: new Map<string, string>(),
    loading: true as boolean
  },
  mutations: {
    [SET_FAVORITE_LANGUAGES](state, payload: {favoriteLanguages: Map<string, string>}) {
      state.favoriteLanguages = payload.favoriteLanguages
    },
    [SET_STATUS_LOAD](state) {
      state.loading = true
    },
    [SET_STATUS_READY](state) {
      state.loading = false
    },
    [CLEAR_FAVORITE_LANGUAGES](state) {
      state.favoriteLanguages = new Map()
    }
  },
  getters: {
    selectedOptions(state): LanguageOption[] {
      return Util.mapToOptions(state.favoriteLanguages)
    }
  },
  actions: {
    async [LOAD_FAVORITE_LANGUAGES]({ commit }) {
      try {
        commit(SET_STATUS_LOAD)
        const { data } = await axios.get<any, AxiosResponse<string[]>>(generateUrl('/apps/ocr/api/personal/languages'))
        const favoriteLanguages = Util.mapToCommonLanguages(data)
        commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages })
        commit(SET_STATUS_READY)
      } catch (e) {
        NextcloudGuiApiService.displayError(Translations.TRANSLATION_UNEXPECTED_ERROR_LOAD)
        commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages: new Map() })
        commit(SET_STATUS_READY)
      }
    },
    async [SAVE_FAVORITE_LANGUAGES]({ commit, dispatch, state }) {
      try {
        commit(SET_STATUS_LOAD)
        const dataToTransfer = { favoriteLanguages: JSON.stringify(Array.from(state.favoriteLanguages.keys())) }
        const { data } = await axios.post<{ favoriteLanguages: string }, AxiosResponse<string[]>>(generateUrl('/apps/ocr/api/personal/languages'), dataToTransfer)
        const favoriteLanguages = Util.mapToCommonLanguages(data)
        commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages })
        commit(SET_STATUS_READY)
      } catch (e) {
        if (e.response.status === 400) {
          NextcloudGuiApiService.displayError(Translations.TRANSLATION_ERROR_WRONG_INPUT)
        } else {
          NextcloudGuiApiService.displayError(Translations.TRANSLATION_UNEXPECTED_ERROR_SAVE)
        }
        await dispatch(LOAD_FAVORITE_LANGUAGES)
      }
    }
  }
})
