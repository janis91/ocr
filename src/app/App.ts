import { TesseractService } from '@a/services/TesseractService'
import { OcaService } from '@a/services/OcaService'
import { Util } from '@a/util/Util'
import { Translations } from '@a/configuration/Translations'
import { NextcloudGuiApiService } from '@a/services/NextcloudGuiApiService'
import { LOAD_FAVORITE_LANGUAGES, SET_SELECTED_FILES, SHOW_MODAL, CLEAR_SELECTED_FILES } from '@a/store/Store'

export class App {
  private tesseractService: TesseractService;
  private ocaService: OcaService;

  constructor() {
    this.ocaService = new OcaService()
    this.tesseractService = new TesseractService()
  }

  public init = () => {
    this.updateFavoriteLanguages()
    this.ocaService.registerFileActions(this.fileActionHandler)
    this.ocaService.registerCheckBoxEvents(this.selectedFilesActionHandler)
    // eslint-disable-next-line no-console
    console.debug('OCA.Ocr initialized')
  }

  public updateFavoriteLanguages = () => {
    OCA.Ocr.store.dispatch(LOAD_FAVORITE_LANGUAGES)
  }

  public process = async (file: OCAFile, selectedLanguages: string[], replace: boolean): Promise<void> => {
    const pdf = await this.tesseractService.process(OCA.Files.App.fileList.getDownloadUrl(file.name), selectedLanguages)
    const newPath = this.ocaService.getCurrentDirectory() + file.name.split('.').filter((k, i) => i !== file.name.split('.').length - 1).join('.') + '.pdf'
    await this.ocaService.putFileContents(newPath, pdf)
    if (replace) {
      await this.ocaService.deleteFile(file.name)
    }
  }

  public fileActionHandler: OCAFileActionHandler = (_something, context) => {
    const selectedFiles = Util.filterFilesWithMimeTypes([context.fileInfoModel.attributes])
    if (selectedFiles.length > 0) {
      OCA.Ocr.store.commit(SET_SELECTED_FILES, { selectedFiles })
      OCA.Ocr.store.commit(SHOW_MODAL)
    } else {
      NextcloudGuiApiService.displayError(`${Translations.TRANSLATION_OCR_PROCESSING_FAILED} ${Translations.TRANSLATION_MIMETYPE_NOT_SUPPORTED}`)
    }
  }

  public selectedFilesActionHandler: () => void = () => {
    const selFiles: Array<OCAFile> = Util.filterFilesWithMimeTypes(OCA.Files.App.fileList.getSelectedFiles())
    if (selFiles.length > 0) {
      this.ocaService.registerMultiSelectMenuItem(() => OCA.Ocr.store.commit(SHOW_MODAL))
      OCA.Ocr.store.commit(SET_SELECTED_FILES, { selectedFiles: selFiles })
    } else {
      this.ocaService.unregisterMultiSelectMenuItem()
      OCA.Ocr.store.commit(CLEAR_SELECTED_FILES)
    }
  }
}
