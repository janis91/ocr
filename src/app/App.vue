<template>
  <div id="ocr-view">
    <modal v-if="show" @close="closeModal" :can-close="canClose" :title="title">
      <modal-content class="modal__content"></modal-content>
    </modal>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Modal } from '@nextcloud/vue/dist/Components/Modal'
import { Translations as CommonTranslations } from '@/common/Translations'
import { HIDE_MODAL } from '@a/store/Store'
import { Translations } from '@a/configuration/Translations'
import ModalContent from '@a/components/ModalContent.vue'

export default Vue.extend({
  name: 'App',
  components: { Modal, ModalContent },
  computed: {
    show(): boolean {
      return this.$store.state.showModal
    },
    title(): string {
      return `${CommonTranslations.TRANSLATION_OCR}: ${Translations.TRANSLATION_FILE_FILES(this.$store.state.selectedFiles.length)}`
    },
    canClose(): boolean {
      return !this.$store.state.processing
    }
  },
  methods: {
    closeModal(): void {
      this.$store.commit(HIDE_MODAL)
    }
  }
})
</script>

<style>
.icon-ocr {
    background-image: url('../img/icon/ocr.svg');
}
.dark .icon-ocr {
    filter: invert(100%);
}
</style>

<style scoped>
.modal__content {
  width: 100%;
  max-width: 900px;
  margin: 15px 15px;
  min-height: 50vh;
  max-height: 80%;
}
</style>
