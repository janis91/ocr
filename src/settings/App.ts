import Vue from 'vue'
import View from '@s/App.vue'
import store from '@s/store/Store'

export class App {
    private view: Vue;

    constructor() {
      this.view = new Vue({
        el: '#ocr-view',
        name: 'OcrView',
        render: h => h(View),
        store
      })
      // eslint-disable-next-line no-console
      console.log('OCA.Ocr initialized')
    }
}
