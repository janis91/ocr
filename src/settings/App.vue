<template>
  <div>
    <multi-select
      :disabled="loading"
      :selectedLanguages="favoriteLanguages"
      @[clear]="clearAll"
      @[update]="updateSelectedLanguages"
      class="select"
    ></multi-select>
    <div class="hint">
      <span class="icon icon-details"></span>
      <span>{{hint}}</span>
    </div>
    <button @click="save" :disabled="loading">{{saveText}}</button>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { LOAD_FAVORITE_LANGUAGES, CLEAR_FAVORITE_LANGUAGES, SET_FAVORITE_LANGUAGES, SAVE_FAVORITE_LANGUAGES } from '@s/store/Store'
import MultiSelect, { MultiSelectEvents } from '@/common/components/MultiSelect.vue'
import { LanguageOption, Util } from '@/common/Util'
import { Translations } from '@s/configuration/Translations'

export default Vue.extend({
  name: 'app',
  components: {
    MultiSelect
  },
  data: () => ({
    clear: MultiSelectEvents.CLEAR_ALL,
    update: MultiSelectEvents.UPDATE_LANGS,
    saveText: Translations.TRANSLATION_SAVE,
    hint: Translations.TRANSLATION_PRESELECTION_HINT
  }),
  computed: {
    favoriteLanguages(): LanguageOption[] {
      return this.$store.getters.selectedOptions
    },
    loading(): boolean {
      return this.$store.state.loading
    }
  },
  methods: {
    clearAll() {
      this.$store.commit(CLEAR_FAVORITE_LANGUAGES)
    },
    updateSelectedLanguages(value: LanguageOption[]) {
      this.$store.commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages: Util.mapOptionsToLanguages(value) })
    },
    save() {
      this.$store.dispatch(SAVE_FAVORITE_LANGUAGES)
    }
  },
  created() {
    this.$store.dispatch(LOAD_FAVORITE_LANGUAGES)
  }
})
</script>

<style scoped>
.select {
  width: 50vw;
  padding-bottom: 10px;
}
.hint {
  padding: 0.5rem 4px;
  color: var(--color-text-lighter);
}

.hint >>> span.icon {
  background-size: 16px 16px;
  display: inline-block;
  vertical-align: text-top;
  margin-right: .3rem;
}
</style>
