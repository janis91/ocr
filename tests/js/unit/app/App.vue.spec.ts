import '../GlobalMocks.mock'
import { shallowMount, Wrapper, createLocalVue } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import App from '@a/App.vue'
import { State, HIDE_MODAL } from '@a/store/Store'
import { FileFixtures } from '../fixtures/FileFixtures'

describe('MultiSelect', () => {
  let wrapper: Wrapper<any>
  let localVue: typeof Vue
  let mutations: any
  let state: Partial<State>

  beforeEach(() => {
    localVue = createLocalVue()
    localVue.use(Vuex)
    state = {
      processing: false,
      selectedFiles: [FileFixtures.PNG],
      showModal: true
    }
    mutations = {
      [HIDE_MODAL]: jest.fn()
    }
    const store = new Vuex.Store({
      strict: true,
      state,
      mutations
    })
    wrapper = shallowMount(App, { store, localVue })
  })

  it('GIVEN standard state, WHEN snapshot is taken, THEN matches snapshot (modal).', async () => {
    expect(wrapper.vm.canClose).toBeTruthy()
    expect(wrapper.vm.show).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('GIVEN state with showModal = false, WHEN snapshot is taken, THEN matches snapshot (no modal).', async () => {
    state.showModal = false
    const store = new Vuex.Store({
      strict: true,
      state,
      mutations
    })
    wrapper = shallowMount(App, { store, localVue })

    expect(wrapper.vm.canClose).toBeTruthy()
    expect(wrapper.vm.show).toBeFalsy()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('GIVEN standard state, WHEN modal emits @close, THEN triggers corresponding mutation.', async () => {
    wrapper.find('modal-stub').vm.$emit('close')

    await wrapper.vm.$nextTick()

    expect(mutations[HIDE_MODAL]).toHaveBeenCalled()
  })
})
