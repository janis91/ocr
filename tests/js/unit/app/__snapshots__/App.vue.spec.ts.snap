// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect GIVEN standard state, WHEN snapshot is taken, THEN matches snapshot (modal). 1`] = `
<div
  id="ocr-view"
>
  <modal-stub
    canclose="true"
    clearviewdelay="5000"
    enableswipe="true"
    size="normal"
    slideshowdelay="3000"
    title="OCR: %n file"
  >
    <modal-content-stub
      class="modal__content"
    />
  </modal-stub>
</div>
`;

exports[`MultiSelect GIVEN state with showModal = false, WHEN snapshot is taken, THEN matches snapshot (no modal). 1`] = `
<div
  id="ocr-view"
>
  <!---->
</div>
`;
